LoRaEngine/LoraKeysManagerFacade/IoTCentralImp/DeviceTwinPageResult.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoraKeysManagerFacade.IoTCentralImp
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net.Http;
+    using System.Threading.Tasks;
+    using LoraKeysManagerFacade.IoTCentralImp.Definitions;
+
+    internal class DeviceTwinPageResult : IRegistryPageResult<IDeviceTwin>
+    {
+        private readonly HttpClient client;
+        private readonly Func<IDeviceTwin, bool> predicate;
+        private readonly string apiVersion;
+
+        private DeviceCollectionResult currentResults;
+
+        public DeviceTwinPageResult(HttpClient client, string apiVersion, Func<IDeviceTwin, bool> predicate)
+        {
+            this.client = client;
+            this.apiVersion = apiVersion;
+            this.predicate = predicate;
+        }
+
+        public bool HasMoreResults => this.currentResults == null || !string.IsNullOrEmpty(this.currentResults.NextLink);
+
+        public async Task<IEnumerable<IDeviceTwin>> GetNextPageAsync()
+        {
+            var results = new List<IDeviceTwin>();
+            if (!this.HasMoreResults)
+            {
+                return results;
+            }
+
+            HttpResponseMessage result = null;
+
+            result = await this.client.GetAsync(new Uri(this.client.BaseAddress,
```

Comment:
I am seeing here that we load the whole devices from IoT Central and then iterate locally to find the matchin one. have you considered using a [query ](https://docs.microsoft.com/en-us/rest/api/iotcentral/1.1-previewdataplane/query/run) 
(e.g. SELECT DevAddr FROM dtmi:modelDefinition:ru5kifjr:ns8lsni8q8 WHERE $id='<id>'" ) to find the correct one immediately and resolve this in IoT central?

Reply:
Hi @Mandur, 

You're right, this is a better option. 
However, I didn't found a way on IoT Central APIs to query the registry.
Please find more info at [https://docs.microsoft.com/en-us/rest/api/iotcentral/](https://docs.microsoft.com/en-us/rest/api/iotcentral/)

Reply:
Hi @Mandur, 

The Azure support team responded that the IoT Central API does not support querying devices in multiple models. 

I created an open question at [https://docs.microsoft.com/en-us/answers/questions/633800/azure-iot-central-query-across-multiple-device-tem.html](https://docs.microsoft.com/en-us/answers/questions/633800/azure-iot-central-query-across-multiple-device-tem.html). 

I hope they can take a look on that and move this question from backlog to their next sprints...

Reply:
Hi @Mandur, 

I pushed an update to use query API rather than looping on devices...

However, since query APIs doesn't giv e the possibility to search accross multiple templates, I realized a loop on device templates that uses the LoRa Device component...

The result will be as many query as device templates that fit the previous request.

I think we can enhance this part with caching the device templates results. During my tests I saw that querying the device templates could take more than 10 secs rather that querying the devices (by /query) runs in less than 1 sec...

Reply:
As IoT Central is using DPS, most likely DPS API should help getting devices, adding, listing them: https://docs.microsoft.com/en-us/rest/api/iot-dps/

Reply:
Hi @Ellbach, 

Thank you for your suggestion. I'll check if querying the DPS by Rest API will do the stuff. But in my mind, it seems that the DPS will not handle device properties (obviously reported properties), so we won't be able to search the device by its DevAddr...

Reply:
Correct, DPS is for the association with the Hub. And looking again at the code, if the device is provisioned "just" with DPS, the code will be similar (once allocated) to what it is with an associated IoT Hub.
Reported properties will show up in the twins, they are just twins. Specific twins but twins.

Reply:
the device will report the twin. Here is an example with a device associated to a modelId, for P&P, so a typical IoT Central device. It does report the properties properly in the twins:
![image](https://user-images.githubusercontent.com/8145578/145594758-83e42d61-5dfc-4bbb-8a65-a2e8e0c0976e.png)

The writable properties are a bit different but you can access them thru PropertyName.value instead of just PropertyName .

Reply:
So most likely, you can simplify quite a lot the way you query the device properties and reuse the non P&P way by adjusting it. That's the only thing specific to P&P and to IoT Central.
DPS will give you the associated hub to query

Reply:
Hi @Ellerbach,

I agree, bu we cannot wuery the IoT Hub for two reasons : 

- We may hav multiple IoT Hub Instances involded behind the IoT Central
- We don't have any key to connect to the IoT Hub and query it...

Reply:
> * We may hav multiple IoT Hub Instances involded behind the IoT Central

That would be ok as you'll get it thru the DPS instance

> * We don't have any key to connect to the IoT Hub and query it...

But that's the real bummer, you're right. Is there a way to pull the keys and have that done somehow in a more transparent way? 

what bother me with your implementation is that it's quite complex while al the elements behind are almost the same as in the original implementation, just that they are split over different hubs and properties are reported a bit differently.

Reply:
> But that's the real bummer, you're right. Is there a way to pull the keys and have that done somehow in a more transparent way ?

I didn't find any way to get the keys.. ;(

> what bother me with your implementation is that it's quite complex while al the elements behind are almost the same as in the original implementation, just that they are split over different hubs and properties are reported a bit differently.

I completly agree with you. To be honest, when starting to connect my devices with IoT Central I guest that it was as simple as creating a switch, adding a configuration and handle the IoT Central with some APIs to get informations about IoT Hub... 
But sincerely, while it's not ideal, since this is a SaaS/AaaS solution, it's quite normal to get only access to the high level.

Reply:
I looked at some of the docs and some of the options. It seems that the operations lists https://docs.microsoft.com/en-us/rest/api/iot-dps/operations/list and then some of the operations allows to get the keys directly of the concerned IoT Hubs. Especially this one: Microsoft.Devices/iotHubs/listkeys/Action
So if that's the case, it can drastically simplify quite some elements.

Reply:
Hi @Ellerbach, 

That's true when working directly with DPS and IoT Hub. 
However, In case of IoT Central, as for IoT Hub, we haven't the keys to the DPS anyway ...